samples/bluetooth/mesh/light_dimmer/src/model_handler.c

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+/**
+ * @file
+ * @brief Model handler for the Light Dimmer Controller.
+ *
+ */
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/mesh/proxy.h>
+#include <bluetooth/mesh/models.h>
+#include <dk_buttons_and_leds.h>
+#include "model_handler.h"
+#include <time.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(dimmer_sample, LOG_LEVEL_INF);
+
+#define SHORT_PRESS_THRESHOLD_MS 300
+#define LONG_PRESS_TIMEOUT_SEC K_SECONDS(20)
+#define DELTA_MOVE_STEP 1000
+
+enum button_state_t { BUTTON_STATE_RELEASED, BUTTON_STATE_PRESSED };
+
+/** Context for a single Light Dimmer instance. */
+static struct dim_button {
+	/** Generic Level client instance for this switch. */
+	struct bt_mesh_lvl_cli client;
+	/* Relative start time of the latest press of the button */
+	int64_t press_start_time;
+	/* Button state: can be "pressed" or "released" */
+	enum button_state_t button_state;
+	/* Toggle state: 0: down/off 1: up/on */
+	bool tgl_state;
+	/* Long press work (Dimmer mode) */
+	struct k_work_delayable long_press_start_work;
+	struct k_work_delayable long_press_stop_work;
+	/* Dimmer mode handlers */
+	void (*const dim_stop_handler)(struct k_work *work);
+	void (*const dim_start_handler)(struct k_work *work);
+};
+
+/* Button 1 handlers */
+static void dim0_start_handler(struct k_work *work);
+static void dim0_stop_handler(struct k_work *work);
+
+static struct dim_button buttons[] = {
+
+	/* Button 1 */
+	{ .client = BT_MESH_LVL_CLI_INIT(NULL),
+	  .press_start_time = 0,
+	  .button_state = BUTTON_STATE_RELEASED,
+	  .tgl_state = false,
+	  .dim_start_handler = dim0_start_handler,
+	  .dim_stop_handler = dim0_stop_handler },
+
+	/* More dimmer buttons can be added here and may be used
+	 * simultaneously.
+	 */
+};
+
+/* Common handlers for dimmer buttons */
+
+static int short_press_handle(struct dim_button dimmer)
```

Comment:
I think this should use a generic onoff client instead? If I understand dim-to-dark behavior correctly, sending BT_MESH_LVL_MIN if the server has any other level than minimum will dim to minimum, not off, which is probably not what you want?